### PR TITLE
[Security] Add API to read frame protector size

### PR DESCRIPTION
**Pull request by gtcooke94**
Adds an API to read frame protector size on the zero copy protector

### CODE REVIEW

File: src/core/tsi/transport_security_grpc.h [hunk 1]
```diff
@@ -64,6 +64,16 @@ void tsi_zero_copy_grpc_protector_destroy(tsi_zero_copy_grpc_protector* self);
 tsi_result tsi_zero_copy_grpc_protector_max_frame_size(
     tsi_zero_copy_grpc_protector* self, size_t* max_frame_size);
 
+// Reads the frame size of the input slice buffer.
+// - protected_slices is the bytes of protected frames.
+// - frame_size is the output frame size.
+// - Returns TSI_OK in case of success.
+// - MUST be called BEFFORE tsi_zero_copy_grpc_protector_unprotect if this value
```
Comment (matthewstevenson88) [inline src/core/tsi/transport_security_grpc.h:71]: What is "this value" referring to? frame_size?

File: src/core/tsi/transport_security_grpc.h [hunk 2]
```diff
@@ -64,6 +64,16 @@ void tsi_zero_copy_grpc_protector_destroy(tsi_zero_copy_grpc_protector* self);
 tsi_result tsi_zero_copy_grpc_protector_max_frame_size(
     tsi_zero_copy_grpc_protector* self, size_t* max_frame_size);
 
+// Reads the frame size of the input slice buffer.
```
Comment (matthewstevenson88) [inline src/core/tsi/transport_security_grpc.h:67]: Can we clarify in the comments that this method does not consume any of the protected_slices and they still need to be fed into unprotect?

In general, I think we need to clarify the relationship more between this method and unprotect. For example, if we call this method once and frame_size returns 1k, then I think calling this method will continue to return 1k until we have unprotected 1k bytes. It would be good to walk through the basic CUJs like this.

Comment (matthewstevenson88) [inline src/core/tsi/transport_security_grpc.h:67]: Similar to the comment on protect/unprotect, please add a line clarifying thread-safety expectations vs. other methods.

File: src/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector.cc [hunk 3]
```diff
@@ -249,21 +249,39 @@ static tsi_result alts_zero_copy_grpc_protector_max_frame_size(
...
+    tsi_zero_copy_grpc_protector* self, grpc_slice_buffer* protected_slices,
+    uint32_t* frame_size) {
+  if (self == nullptr || frame_size == nullptr) return TSI_INVALID_ARGUMENT;
+  alts_zero_copy_grpc_protector* protector =
+      reinterpret_cast<alts_zero_copy_grpc_protector*>(self);
+
+  if (protector->parsed_frame_size == 0) {
+    // We have not parsed frame size yet. Parses frame size.
+    if (!read_frame_size(protected_slices, &protector->parsed_frame_size)) {
+      return TSI_DATA_CORRUPTED;
```
Comment (matthewstevenson88) [inline src/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector.cc:262]: I'm not sure this is what we want to do: if `protected_slices` contains only the first half of the header, then we are going to return data corrupted, which will cause the connection to get closed.

File: test/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector_test.cc [hunk 4]
```diff
@@ -194,6 +195,11 @@ static void seal_unseal_small_buffer(tsi_zero_copy_grpc_protector* sender,
         1;
     grpc_slice_buffer_move_first(&var->protected_sb, staging_sb_size,
                                  &var->staging_sb);
+    uint32_t frame_size;
+    ASSERT_EQ(tsi_zero_copy_grpc_protector_read_frame_size(
```
Comment (matthewstevenson88) [inline test/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector_test.cc:199]: Can we please add tests for more edge cases?

**Review by matthewstevenson88 — COMMENTED** [review]
Thanks for the PR Greg!

File: src/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector.cc [hunk 5]
```diff
@@ -249,21 +249,39 @@ static tsi_result alts_zero_copy_grpc_protector_max_frame_size(
   return TSI_OK;
 }
 
+static bool alts_zero_copy_grpc_protector_read_frame_size(
+    tsi_zero_copy_grpc_protector* self, grpc_slice_buffer* protected_slices,
+    uint32_t* frame_size) {
+  if (self == nullptr || frame_size == nullptr) return TSI_INVALID_ARGUMENT;
```
Comment (ananda1066) [inline src/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector.cc:255]: should this function return a tsi_result instead of a bool if we return TSI_INVALID_ARGUMENT here (and to stay consistent with the rest of the API)?

File: src/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector.cc [hunk 6]
```diff
@@ -249,21 +249,39 @@ static tsi_result alts_zero_copy_grpc_protector_max_frame_size(
...
+  alts_zero_copy_grpc_protector* protector =
+      reinterpret_cast<alts_zero_copy_grpc_protector*>(self);
+
+  if (protector->parsed_frame_size == 0) {
+    // We have not parsed frame size yet. Parses frame size.
+    if (!read_frame_size(protected_slices, &protector->parsed_frame_size)) {
+      return false;
```
Comment (ananda1066) [inline src/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector.cc:262]: This condition is different than in unprotect - in unprotect, if read_frame_size is 0, it returns TSI_DATA_CORRUPTED since the code already made sure there was enough to read to get frame size. Here, false could mean the frame size could not be read or we don't have enough bytes yet to read frame size. Would we want to return an error if the frame size could not be read?

I'm wondering if it would be simpler to add a frame_size field to unprotect that is set similar to min_progress_size to avoid mutating protected_slices in two places / having inconsistent errors?

File: src/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector.cc [hunk 7]
```diff
@@ -249,21 +249,29 @@ static tsi_result alts_zero_copy_grpc_protector_max_frame_size(
   return TSI_OK;
 }
 
+static bool alts_zero_copy_grpc_protector_read_frame_size(
+    tsi_zero_copy_grpc_protector*, grpc_slice_buffer* protected_slices,
+    uint32_t* frame_size) {
+  if (frame_size == nullptr) return TSI_INVALID_ARGUMENT;
```
Comment (matthewstevenson88) [inline src/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector.cc]: Should we be returning a bool rather than TSI_INVALID_ARGUMENT?

File: src/core/tsi/fake_transport_security.cc [hunk 8]
```diff
@@ -520,12 +520,32 @@ static tsi_result fake_zero_copy_grpc_protector_max_frame_size(
   return TSI_OK;
 }
 
+static bool fake_zero_copy_grpc_protector_read_frame_size(
+    tsi_zero_copy_grpc_protector* self, grpc_slice_buffer* protected_slices,
+    uint32_t* frame_size) {
+  if (self == nullptr || frame_size == nullptr) return TSI_INVALID_ARGUMENT;
+  tsi_fake_zero_copy_grpc_protector* impl =
+      reinterpret_cast<tsi_fake_zero_copy_grpc_protector*>(self);
+  while (protected_slices->length >= TSI_FAKE_FRAME_HEADER_SIZE) {
```
Comment (matthewstevenson88) [inline src/core/tsi/fake_transport_security.cc:528]: Should we also change the fake implementation to be stateless?

File: src/core/tsi/transport_security_grpc.h [hunk 9]
```diff
@@ -64,6 +64,23 @@ void tsi_zero_copy_grpc_protector_destroy(tsi_zero_copy_grpc_protector* self);
 tsi_result tsi_zero_copy_grpc_protector_max_frame_size(
     tsi_zero_copy_grpc_protector* self, size_t* max_frame_size);
 
+// In general, I think we need to clarify the relationship more between this
```
Comment (matthewstevenson88) [inline src/core/tsi/transport_security_grpc.h]: Should we remove this comment? This seems like a note to self that maybe we forgot to remove?

File: src/core/tsi/transport_security_grpc.h [hunk 10]
```diff
@@ -64,6 +64,23 @@ void tsi_zero_copy_grpc_protector_destroy(tsi_zero_copy_grpc_protector* self);
...
+// method and unprotect. For example, if we call this method once and
+// frame_size returns 1k, then I think calling this method will continue to
+// return 1k until we have unprotected 1k bytes. It would be good to walk
+// through the basic CUJs like this.
+
+// A stateless methods that reads the frame size of the input slice buffer. Does
+// NOT consume any of the protected_slices does NOT modify the `self` input
```
Comment (matthewstevenson88) [inline src/core/tsi/transport_security_grpc.h]: nit: s/does/and does

File: test/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector_test.cc [hunk 11]
```diff
@@ -37,6 +37,8 @@ constexpr size_t kSmallBufferSize = 16;
 constexpr size_t kLargeBufferSize = 16384;
 constexpr size_t kChannelMaxSize = 2048;
 constexpr size_t kChannelMinSize = 128;
+constexpr size_t kOverhead = 24;
```
Comment (matthewstevenson88) [inline test/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector_test.cc]: nit: Can we call this variable kFrameOverhead and clarify in a comment that it takes into account both the 8-byte header and the 16-byte authentication tag?

File: test/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector_test.cc [hunk 12]
```diff
@@ -194,6 +197,13 @@ static void seal_unseal_small_buffer(tsi_zero_copy_grpc_protector* sender,
         1;
     grpc_slice_buffer_move_first(&var->protected_sb, staging_sb_size,
                                  &var->staging_sb);
+    // TODO(gtcooke94) call the test with and without this
+    if (do_frame_size_read) {
```
Comment (matthewstevenson88) [inline test/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector_test.cc]: Do we need the do_frame_size_read boolean or the TODO now that the function is stateless? Can we always just call the read frame size API?

File: test/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector_test.cc [hunk 13]
```diff
@@ -232,17 +243,75 @@ static void seal_unseal_large_buffer(tsi_zero_copy_grpc_protector* sender,
     uint32_t channel_size = gsec_test_bias_random_uint32(static_cast<uint32_t>(
                                 kChannelMaxSize + 1 - kChannelMinSize)) +
                             static_cast<uint32_t>(kChannelMinSize);
-    while (var->protected_sb.length > channel_size) {
+    if (do_frame_size_read) {
```
Comment (matthewstevenson88) [inline test/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector_test.cc]: Do we still need this test to be so complex now that the API is stateless?

File: test/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector_test.cc [hunk 14]
```diff
@@ -320,6 +421,75 @@ TEST(AltsZeroCopyGrpcProtectorTest, MainTest) {
   grpc_shutdown();
 }
 
+TEST(AltsZeroCopyFrameProtectorTest, ReadFrameSizeSuccesSmall) {
```
Comment (matthewstevenson88) [inline test/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector_test.cc]: nit: typo in "Succes", similar for the other tests below

File: src/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector.cc (hunk 7 above)
Comment (gtcooke94) [reply src/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector.cc]: Yes, my bad

File: src/core/tsi/fake_transport_security.cc (hunk 8 above)
Comment (gtcooke94) [reply src/core/tsi/fake_transport_security.cc:528]: Fixed

File: src/core/tsi/transport_security_grpc.h (hunk 9 above)
Comment (gtcooke94) [reply src/core/tsi/transport_security_grpc.h]: My bad

File: test/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector_test.cc (hunk 12 above)
Comment (gtcooke94) [reply test/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector_test.cc]: In the small buffer case I think we can always call it

File: test/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector_test.cc (hunk 13 above)
Comment (gtcooke94) [reply test/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector_test.cc]: So, my intention with this test was to show how the read_frame_size and unprotect can go together - essentially, the bool switch here on `do_frame_size_read` runs the old test on the `else`, then on the `do_frame_size_read` it manually builds the buffers for unprotect instead.

I'm happy to remove it and/or move it elsewhere, WDYT

File: test/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector_test.cc (hunk 14 above)
Comment (gtcooke94) [reply test/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector_test.cc]: Fixed

File: test/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector_test.cc (hunk 13 above)
Comment (gtcooke94) [reply test/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector_test.cc]: Thinking about it more, removed it. Have saved the example elsewhere, but it doesn't belong in a test like this.

Comment (matthewstevenson88) [reply test/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector_test.cc]: Agreed. Thanks!

**Review by matthewstevenson88 — APPROVED** [review]
